addons/mass_mailing/models/mailing.py

```diff
@@ -664,12 +664,18 @@ def action_cancel(self):
         self.write({'state': 'draft', 'schedule_date': False, 'schedule_type': 'now', 'next_departure': False})
 
     def action_retry_failed(self):
-        failed_mails = self.env['mail.mail'].sudo().search([
-            ('mailing_id', 'in', self.ids),
-            ('state', '=', 'exception')
-        ])
-        failed_mails.mapped('mailing_trace_ids').unlink()
-        failed_mails.unlink()
+        """ Remove all failed emails and their traces, and try sending them again."""
+        batch_size = 1000
+        while True:
```

Comment:
Let's just justify it then
```suggestion
       # batch to overfilling the cache if read is performed during unlink
        batch_size = 1000
        while True:
```